File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
//프로그래머스 올바른 괄호
class Solution {
	boolean solution(String s) {
		int cnt = 0;

		System.out.println(s.charAt(0));
		if (s.charAt(0) == ')' || s.charAt(s.length() - 1) == '(') {
			return false;
		}

		for (int i = 0; i < s.length(); i++) {
			if (s.charAt(i) == '(') {
				cnt++;
			} else if (s.charAt(i) == ')') {
				cnt--;
			}

			if (cnt < 0) {
				return false;
			}
		}

		if (cnt < 0 || cnt > 0) {
			return false;
		}
		return cnt == 0;
	}
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
// 올바른 괄호(스택)
// 스택은 따로없고 배열의 push/pop을 이용
// 큐는 따로없고 배열의 push/shift를 이용
function solution(s) {
    let stack = [];

    for (let i = 0; i < s.length; i++) {
        if (stack.length == 0) {
            stack.push(s[i]);
        } else if (s[i] == ')' && stack[0] == '(') {
            stack.pop();
        } else {
            stack.push(s[i]);
        }
    }
    // 비었다면 참, 아니라면 거짓
    if (stack.length == 0) return true;
    return false;
}